keras_cv/layers/preprocessing3d/random_copy_paste.py

```diff
@@ -0,0 +1,257 @@
+            )
+            object_bounding_boxes = tf.boolean_mask(
+                object_bounding_boxes, object_mask, axis=1
+            )
+        shuffle_index = tf.range(tf.shape(object_point_clouds)[1])
+        shuffle_index = tf.random.shuffle(shuffle_index)
+        object_point_clouds = tf.gather(object_point_clouds, shuffle_index, axis=1)
+        object_bounding_boxes = tf.gather(object_bounding_boxes, shuffle_index, axis=1)
+
+        # Load at most 5 times num_paste_bounding_boxes to check overlaps.
+        num_compare_bounding_boxes = tf.math.minimum(
+            num_paste_bounding_boxes * 5,
+            tf.shape(object_point_clouds)[1],
+        )
+
+        object_point_clouds = object_point_clouds[:, :num_compare_bounding_boxes, :]
+        object_bounding_boxes = object_bounding_boxes[:, :num_compare_bounding_boxes, :]
+        # Use the current frame to check overlap between existing bounding boxes and pasted bounding boxes
+        all_bounding_boxes = tf.concat([bounding_boxes, object_bounding_boxes], axis=1)[
+            0, :, :7
+        ]
+        iou = iou_3d(all_bounding_boxes, all_bounding_boxes)
+        iou = tf.linalg.band_part(iou, -1, 0)
+        iou_sum = tf.reduce_sum(iou[num_existing_bounding_boxes:], axis=1)
+        # A non overlapping bounding box has a 1.0 IoU with itself.
+        non_overlapping_mask = tf.reshape(iou_sum <= 1, [-1])
+        object_point_clouds = tf.boolean_mask(
+            object_point_clouds, non_overlapping_mask, axis=1
+        )
+        object_bounding_boxes = tf.boolean_mask(
+            object_bounding_boxes, non_overlapping_mask, axis=1
+        )
+        object_point_clouds = object_point_clouds[:, :num_paste_bounding_boxes, :]
+        object_bounding_boxes = object_bounding_boxes[:, :num_paste_bounding_boxes, :]
+        return {
+            OBJECT_POINT_CLOUDS: object_point_clouds,
+            OBJECT_BOUNDING_BOXES: object_bounding_boxes,
+        }
+
+    def augment_point_clouds_bounding_boxes(
+        self, point_clouds, bounding_boxes, transformation, **kwargs
+    ):
+        additional_object_point_clouds = transformation[OBJECT_POINT_CLOUDS]
+        additional_object_bounding_boxes = transformation[OBJECT_BOUNDING_BOXES]
+        original_point_clouds_shape = point_clouds.get_shape().as_list()
+        original_object_bounding_boxes = bounding_boxes.get_shape().as_list()
+        points_in_paste_bounding_boxes = tf.reduce_any(
+            is_within_box3d(
+                point_clouds[:, :, :3], additional_object_bounding_boxes[:, :, :7]
+            ),
+            axis=-1,
+        )
+        num_frames = point_clouds.get_shape().as_list()[0]
+        point_clouds_list = []
+        bounding_boxes_list = []
+        for frame_index in range(num_frames):
+            # Remove background point clouds that are in object_bounding_boxes.
+            existing_point_clouds_mask = ~points_in_paste_bounding_boxes[
+                frame_index, :
```

Comment:
In Line 120, iou3d is computed for only 1st frame. But here within_box_3d is computed for each frame. Is there a reason for that? Maybe document it?

Reply:
iou3d is to check overlap between boxes.
within_box_3d is to check and remove background points clouds overlaps with pasted boxes.